### PR TITLE
Remove unused inputs for some operators

### PR DESCRIPTION
**Pull request by zhiqiu**
In #21169, we add put some operators that cannot pass `unused-input-check` to `white_list`.
This PR reduces the number of operators in `white_list` from 34 to 17, by carefully removing unused inputs for some operators.

### CODE REVIEW

File: paddle/fluid/framework/unused_var_check.cc [hunk 1]
```diff
@@ -27,40 +27,24 @@ DEFINE_bool(enable_unused_var_check, false,
             "especially for grad operator. It should be in unittest.");
 
 const std::unordered_set<std::string> op_has_unsed_vars_white_list = {
-    "auc",
     "batch_norm",
     "batch_norm_grad",
```
Comment (luotao1) [inline paddle/fluid/framework/unused_var_check.cc]: 以下三种情况不能修复，需要继续保留在白名单中。
- 分支，不同分支用不同的input，
- fused op，
- 使用input的dtype信息

请在白名单里填上注释，并注明每个Op属于哪一种情况。可以在下一个PR完成。

File: paddle/fluid/operators/center_loss_op.cc [hunk 2]
```diff
@@ -141,6 +141,9 @@ class CenterLossOpGradMaker : public framework::SingleGradOpMaker<T> {
     return retv;
   }
 };
+
+DECLARE_NO_NEED_BUFFER_VARS_INFERENCE(CenterLossGradNoNeedBufVarsInferer, "X");
```
Comment (sneaxiy) [inline paddle/fluid/operators/center_loss_op.cc:145]: Seems that the CUDA kernel of `center_loss_grad` would use the buffer of `X`, see `center_loss_op.cu` to check it? Maybe we should change the implementation of the CUDA kernel.

File: paddle/fluid/operators/crop_op.cc [hunk 3]
```diff
@@ -201,13 +201,16 @@ class CropGradOpMaker : public framework::SingleGradOpMaker<T> {
   }
 };
 
+DECLARE_NO_NEED_BUFFER_VARS_INFERENCE(GropNoNeedBufferVarInference, "Y");
```
Comment (sneaxiy) [inline paddle/fluid/operators/crop_op.cc:204]: Seems that `X` is also not needed? Please see the implementation of this op. "X" can be removed from the grad op maker.

File: paddle/fluid/operators/gru_op.cc [hunk 4]
```diff
@@ -376,15 +384,53 @@ class GRUCPUKernel : public framework::OpKernel<T> {
   }
 };
 
+template <typename T>
+class CRUGradOpMaker : public framework::SingleGradOpMaker<T> {
```
Comment (sneaxiy) [inline paddle/fluid/operators/gru_op.cc]: CRU -> GRU.

File: paddle/fluid/operators/gru_op.cc [hunk 5]
```diff
@@ -376,15 +384,53 @@ class GRUCPUKernel : public framework::OpKernel<T> {
...
 
 namespace ops = paddle::operators;
-REGISTER_OPERATOR(
-    gru, ops::GRUOp, ops::GRUOpMaker,
-    paddle::framework::DefaultGradOpMaker<paddle::framework::OpDesc, true>,
-    paddle::framework::DefaultGradOpMaker<paddle::imperative::OpBase, true>)
-REGISTER_OPERATOR(gru_grad, ops::GRUGradOp);
+REGISTER_OPERATOR(gru, ops::GRUOp, ops::GRUOpMaker,
+                  ops::CRUGradOpMaker<paddle::framework::OpDesc>,
+                  ops::CRUGradOpMaker<paddle::imperative::OpBase>)
```
Comment (sneaxiy) [inline paddle/fluid/operators/gru_op.cc]: Add `;` to the end for better coding style.

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 6]
```diff
@@ -313,15 +314,36 @@ class CPUMatchMatrixTensorOPGradKernel : public framework::OpKernel<T> {
...
 namespace ops = paddle::operators;
-REGISTER_OPERATOR(
-    match_matrix_tensor, ops::MatchMatrixTensorOP,
-    ops::MatchMatrixTensorOpMaker,
-    paddle::framework::DefaultGradOpMaker<paddle::framework::OpDesc, true>,
-    paddle::framework::DefaultGradOpMaker<paddle::imperative::OpBase, true>)
+REGISTER_OPERATOR(match_matrix_tensor, ops::MatchMatrixTensorOP,
+                  ops::MatchMatrixTensorOpMaker,
+                  ops::MatchMatrixTensorGradOpMaker<paddle::framework::OpDesc>,
+                  ops::MatchMatrixTensorGradOpMaker<paddle::imperative::OpBase>)
```
Comment (sneaxiy) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: Add `;` at the end.

File: paddle/fluid/operators/center_loss_op.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/operators/center_loss_op.cc]: Thanks for finding this, I will add center_loss_grad to the white_list and request OP owner to change it later.

File: paddle/fluid/operators/crop_op.cc (hunk 3 above)
Comment (zhiqiu) [reply paddle/fluid/operators/crop_op.cc:204]: I find `X` used here. https://github.com/PaddlePaddle/Paddle/blob/36afdbd3e11149208ce4d779e4171d4e2c054611/paddle/fluid/operators/crop_op.h#L62-L63

File: paddle/fluid/framework/unused_var_check.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/framework/unused_var_check.cc]: 好的，我会添加注释说明。

File: paddle/fluid/operators/center_loss_op.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/operators/center_loss_op.cc]: It seems that OpTest cannot check the correctness of `DECLARE_NO_NEED_BUFFER_VARS_INFERENCE`? Maybe we need to enhance that?

File: paddle/fluid/operators/gru_op.cc (hunk 4 above)
Comment (zhiqiu) [reply paddle/fluid/operators/gru_op.cc]: Done, thanks.

File: paddle/fluid/operators/gru_op.cc (hunk 5 above)
Comment (zhiqiu) [reply paddle/fluid/operators/gru_op.cc]: Done, thanks.

File: paddle/fluid/operators/match_matrix_tensor_op.cc (hunk 6 above)
Comment (zhiqiu) [reply paddle/fluid/operators/match_matrix_tensor_op.cc]: Done, thanks.

**Review by luotao1 — DISMISSED** [review]
LGTM! @sneaxiy Please have a double review!

File: paddle/fluid/operators/center_loss_op.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/operators/center_loss_op.cc:145]: My fault. This changing is right.

File: paddle/fluid/operators/crop_op.cc (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/operators/crop_op.cc:204]: My fault. This changing is right.

**Review by sneaxiy — APPROVED** [review]

**Pull request merged**